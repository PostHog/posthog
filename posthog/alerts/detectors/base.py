"""Base classes and types for alert detectors."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import StrEnum
from typing import Any, Optional


class DetectorType(StrEnum):
    """Available detector types."""

    THRESHOLD = "threshold"
    ZSCORE = "zscore"
    MAD = "mad"  # Median Absolute Deviation


class DetectionDirection(StrEnum):
    """Direction for anomaly detection."""

    UP = "up"  # Alert on values above normal
    DOWN = "down"  # Alert on values below normal
    BOTH = "both"  # Alert on values above or below normal


class ValueType(StrEnum):
    """Whether to run detection on raw values or deltas."""

    RAW = "raw"  # Run detection on raw metric values
    DELTA = "delta"  # Run detection on period-over-period changes


@dataclass
class DetectionResult:
    """Result of running a detector on data."""

    value: Optional[float]  # The raw metric value
    detector_score: Optional[float]  # The detector-specific score (e.g., z-score, mad score)
    is_breach: bool  # Whether this constitutes a breach
    breach_messages: list[str]  # Human-readable breach descriptions
    metadata: dict[str, Any]  # Additional detector-specific info


class BaseDetector(ABC):
    """Base class for all alert detectors."""

    def __init__(self, config: dict[str, Any]):
        """Initialize detector with configuration."""
        self.config = config
        self.validate_config()

    @abstractmethod
    def detect(
        self, values: list[float], series_name: str = "Series", value_type: ValueType = ValueType.RAW
    ) -> DetectionResult:
        """
        Run detection on a list of values.

        Args:
            values: Time series values (most recent last)
            series_name: Name of the series for breach messages
            value_type: Whether to use raw values or deltas

        Returns:
            DetectionResult with breach status and details
        """
        pass

    @abstractmethod
    def validate_config(self) -> None:
        """Validate the detector configuration. Raise ValueError if invalid."""
        pass

    @property
    @abstractmethod
    def detector_type(self) -> DetectorType:
        """Return the type of this detector."""
        pass

    def _compute_deltas(self, values: list[float]) -> list[float]:
        """Compute period-over-period deltas from values."""
        if len(values) < 2:
            return []
        return [values[i] - values[i - 1] for i in range(1, len(values))]
