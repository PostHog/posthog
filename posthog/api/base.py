from rest_framework.pagination import CursorPagination as RestCursorPagination


class CursorPagination(RestCursorPagination):
    ordering = "-created_at"
    page_size = 100
