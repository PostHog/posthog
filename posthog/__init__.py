import sys
from unittest.mock import MagicMock
import deltalake.writer

# DLT imports the missing function `try_get_deltatable` from `deltalake` which
# it doesn't actually use but also no longer exists. So we can safely mock
# it to allow the app to startup without module errors
deltalake.writer.try_get_deltatable = MagicMock()  # type: ignore
sys.modules["deltalake.writer"] = deltalake.writer


# This will make sure the app is always imported when
# Django starts so that shared_task will use this app.
from posthog.celery import app as celery_app  # noqa: E402

__all__ = ("celery_app",)

# snowflake-connector-python tries to access a root folder which errors out in pods.
# This sets the snowflake home directory to a relative folder
import os  # noqa: E402

os.environ["SNOWFLAKE_HOME"] = "./.snowflake"
