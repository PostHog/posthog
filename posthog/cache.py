from typing import Optional

import fakeredis  # type: ignore
import redis
from django.conf import settings

redis_instance = None  # type: Optional[redis.Redis]

if settings.TEST:
    redis_instance = fakeredis.FakeStrictRedis()
elif settings.REDIS_URL:
    redis_instance = redis.from_url(settings.REDIS_URL, db=0)


def get_cache_key(team_id: int, key: str) -> str:
    return "@c/{}/{}".format(team_id, key)


def get_cached_value(team_id: int, key: str) -> Optional[str]:
    if not redis_instance:
        raise Exception("Redis not configured!")
    return redis_instance.get(get_cache_key(team_id, key))


def set_cached_value(team_id: int, key: str, value: str) -> None:
    if not redis_instance:
        raise Exception("Redis not configured!")
    redis_instance.set(get_cache_key(team_id, key), value)


def clear_cache() -> None:
    if not settings.TEST and not settings.DEBUG:
        raise Exception("Can only clear redis cache in TEST or DEBUG mode!")
    if not redis_instance:
        raise Exception("Redis not configured!")

    redis_instance.flushdb()
