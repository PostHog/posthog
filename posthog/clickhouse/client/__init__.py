from posthog.clickhouse.client.execute import query_with_columns, sync_execute
from posthog.clickhouse.client.execute_async import execute_with_progress

__all__ = [
    "sync_execute",
    "query_with_columns",
    "execute_with_progress",
]
