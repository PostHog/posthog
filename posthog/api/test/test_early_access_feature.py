from unittest.mock import ANY

from rest_framework import status
from django.core.cache import cache
from django.test.client import Client

from posthog.models.early_access_feature import EarlyAccessFeature
from posthog.models import FeatureFlag, Person
from posthog.test.base import (
    APIBaseTest,
    BaseTest,
    QueryMatchingTest,
    snapshot_postgres_queries,
)


class TestEarlyAccessFeature(APIBaseTest):
    maxDiff = None

    def test_can_create_early_access_feature(self):
        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "concept",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data
        assert EarlyAccessFeature.objects.filter(id=response_data["id"]).exists()
        assert FeatureFlag.objects.filter(key=response_data["feature_flag"]["key"]).exists()
        assert response_data["name"] == "Hick bondoogling"
        assert response_data["description"] == 'Boondoogle your hicks with one click. Just click "bazinga"!'
        assert response_data["stage"] == "concept"
        assert response_data["feature_flag"]["key"] == "hick-bondoogling"
        assert response_data["feature_flag"]["active"]
        assert not response_data["feature_flag"]["filters"].get("super_groups", None)
        assert len(response_data["feature_flag"]["filters"]["groups"]) == 1
        assert response_data["feature_flag"]["filters"]["groups"][0]["rollout_percentage"] == 0
        assert isinstance(response_data["created_at"], str)

    def test_promote_to_beta(self):
        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "concept",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data

        feature_id = response_data["id"]

        response = self.client.patch(
            f"/api/projects/{self.team.id}/early_access_feature/{feature_id}",
            data={
                "stage": EarlyAccessFeature.Stage.BETA,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_200_OK, response_data
        assert response_data["stage"] == EarlyAccessFeature.Stage.BETA
        assert len(response_data["feature_flag"]["filters"]["super_groups"]) == 1

    def test_archive(self):
        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data
        assert len(response_data["feature_flag"]["filters"]["super_groups"]) == 1

        feature_id = response_data["id"]

        response = self.client.patch(
            f"/api/projects/{self.team.id}/early_access_feature/{feature_id}",
            data={
                "stage": EarlyAccessFeature.Stage.ARCHIVED,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_200_OK, response_data
        assert response_data["stage"] == EarlyAccessFeature.Stage.ARCHIVED
        assert not response_data["feature_flag"]["filters"].get("super_groups", None)

    def test_update_doesnt_remove_super_condition(self):
        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data
        assert len(response_data["feature_flag"]["filters"]["super_groups"]) == 1

        feature_id = response_data["id"]

        response = self.client.patch(
            f"/api/projects/{self.team.id}/early_access_feature/{feature_id}",
            data={
                "description": "Something else!",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_200_OK, response_data
        assert response_data["stage"] == EarlyAccessFeature.Stage.BETA
        assert response_data["description"] == "Something else!"
        assert len(response_data["feature_flag"]["filters"]["super_groups"]) == 1

    def test_we_dont_delete_existing_flag_information_when_creating_early_access_feature(self):
        flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [
                    {
                        "properties": [{"key": "xyz", "value": "ok", "type": "person"}],
                        "rollout_percentage": None,
                    }
                ],
                "payloads": {"true": "Hick bondoogling? ????"},
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data
        assert EarlyAccessFeature.objects.filter(id=response_data["id"]).exists()
        assert FeatureFlag.objects.filter(key=response_data["feature_flag"]["key"]).exists()

        flag.refresh_from_db()
        self.assertEqual(
            flag.filters,
            {
                "groups": [
                    {
                        "properties": [{"key": "xyz", "value": "ok", "type": "person"}],
                        "rollout_percentage": None,
                    }
                ],
                "payloads": {"true": "Hick bondoogling? ????"},
                "super_groups": [
                    {
                        "properties": [
                            {
                                "key": "$feature_enrollment/hick-bondoogling",
                                "operator": "exact",
                                "type": "person",
                                "value": ["true"],
                            }
                        ],
                        "rollout_percentage": 100,
                    }
                ],
            },
        )

    def test_cant_create_early_access_feature_with_duplicate_key(self):
        FeatureFlag.objects.create(
            team=self.team,
            filters={"groups": [{"properties": [], "rollout_percentage": None}]},
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_400_BAD_REQUEST, response_data

        self.assertEqual(
            response_data["detail"],
            "There is already a feature flag with this key.",
        )

    def test_can_create_new_early_access_feature_with_soft_deleted_flag(self):
        FeatureFlag.objects.create(
            team=self.team,
            filters={"groups": [{"properties": [], "rollout_percentage": None}]},
            key="hick-bondoogling",
            created_by=self.user,
            deleted=True,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data

        assert response.status_code == status.HTTP_201_CREATED, response_data
        assert EarlyAccessFeature.objects.filter(id=response_data["id"]).exists()
        assert FeatureFlag.objects.filter(key=response_data["feature_flag"]["key"]).exists()
        assert response_data["name"] == "Hick bondoogling"
        assert response_data["description"] == 'Boondoogle your hicks with one click. Just click "bazinga"!'
        assert response_data["stage"] == "beta"
        assert response_data["feature_flag"]["key"] == "hick-bondoogling"
        assert response_data["feature_flag"]["active"]
        assert len(response_data["feature_flag"]["filters"]["super_groups"]) == 1
        assert len(response_data["feature_flag"]["filters"]["groups"]) == 1
        assert response_data["feature_flag"]["filters"]["groups"][0]["rollout_percentage"] == 0
        assert isinstance(response_data["created_at"], str)

    def test_deleting_early_access_feature_removes_super_condition_from_flag(self):
        existing_flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [
                    {
                        "properties": [{"key": "xyz", "value": "ok", "type": "person"}],
                        "rollout_percentage": None,
                    }
                ]
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": existing_flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data

        response = self.client.delete(
            f"/api/projects/{self.team.id}/early_access_feature/{response_data['id']}/",
            format="json",
        )

        assert response.status_code == status.HTTP_204_NO_CONTENT
        flag = FeatureFlag.objects.filter(key=response_data["feature_flag"]["key"]).all()[0]

        self.assertEqual(
            flag.filters,
            {
                "groups": [
                    {
                        "properties": [{"key": "xyz", "value": "ok", "type": "person"}],
                        "rollout_percentage": None,
                    }
                ],
                "super_groups": None,
            },
        )

    def test_cant_soft_delete_flag_with_early_access_feature(self):
        existing_flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [
                    {
                        "properties": [{"key": "xyz", "value": "ok", "type": "person"}],
                        "rollout_percentage": None,
                    }
                ]
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": existing_flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_201_CREATED, response_data

        response = self.client.patch(
            f"/api/projects/{self.team.id}/feature_flags/{existing_flag.id}/",
            data={
                "deleted": True,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_400_BAD_REQUEST, response_data

        assert (
            response_data["detail"]
            == "Cannot delete a feature flag that is in use with early access features. Please delete the early access feature before deleting the flag."
        )

    def test_cant_create_early_access_feature_with_group_flag(self):
        flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [{"properties": [], "rollout_percentage": None}],
                "aggregation_group_type_index": 1,
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_400_BAD_REQUEST, response_data

        self.assertEqual(
            response_data["detail"],
            "Group-based feature flags are not supported for Early Access Features.",
        )

    def test_cant_create_early_access_feature_with_multivariate_flag(self):
        flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [{"properties": [], "rollout_percentage": None}],
                "multivariate": {
                    "variants": [
                        {
                            "key": "first-variant",
                            "name": "First Variant",
                            "rollout_percentage": 50,
                        },
                        {
                            "key": "second-variant",
                            "name": "Second Variant",
                            "rollout_percentage": 25,
                        },
                        {
                            "key": "third-variant",
                            "name": "Third Variant",
                            "rollout_percentage": 25,
                        },
                    ]
                },
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_400_BAD_REQUEST, response_data

        self.assertEqual(
            response_data["detail"],
            "Multivariate feature flags are not supported for Early Access Features.",
        )

    def test_cant_create_early_access_feature_with_flag_with_existing_early_access_feature(self):
        flag = FeatureFlag.objects.create(
            team=self.team,
            filters={
                "groups": [{"properties": [], "rollout_percentage": None}],
            },
            key="hick-bondoogling",
            created_by=self.user,
        )

        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Hick bondoogling",
                "description": 'Boondoogle your hicks with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": flag.id,
            },
            format="json",
        )

        # Request for new feature with same flag id should fail
        response = self.client.post(
            f"/api/projects/{self.team.id}/early_access_feature/",
            data={
                "name": "Another feature",
                "description": 'Boondoogle your hicks AGAIN with one click. Just click "bazinga"!',
                "stage": "beta",
                "feature_flag_id": flag.id,
            },
            format="json",
        )
        response_data = response.json()

        assert response.status_code == status.HTTP_400_BAD_REQUEST, response_data

        self.assertEqual(
            response_data["detail"],
            "Linked feature flag hick-bondoogling already has a feature attached to it.",
        )

    def test_can_edit_feature(self):
        feature = EarlyAccessFeature.objects.create(
            team=self.team,
            name="Click counter",
            description="A revolution in usability research: now you can count clicks!",
            stage="beta",
        )

        response = self.client.patch(
            f"/api/projects/{self.team.id}/early_access_feature/{feature.id}",
            data={
                "name": "Mouse-up counter",
                "description": "Oops, we made a mistake, it actually only counts mouse-up events.",
            },
            format="json",
        )
        response_data = response.json()

        feature.refresh_from_db()
        assert response.status_code == status.HTTP_200_OK, response_data
        assert response_data["name"] == "Mouse-up counter"
        assert response_data["description"] == "Oops, we made a mistake, it actually only counts mouse-up events."
        assert response_data["stage"] == "beta"
        assert feature.name == "Mouse-up counter"

    def test_can_list_features(self):
        EarlyAccessFeature.objects.create(
            team=self.team,
            name="Click counter",
            description="A revolution in usability research: now you can count clicks!",
            stage="beta",
        )

        response = self.client.get(f"/api/projects/{self.team.id}/early_access_feature/")
        response_data = response.json()

        assert response.status_code == status.HTTP_200_OK, response_data
        assert response_data == {
            "count": 1,
            "next": None,
            "previous": None,
            "results": [
                {
                    "created_at": ANY,
                    "description": "A revolution in usability research: now you can count clicks!",
                    "documentation_url": "",
                    "feature_flag": None,
                    "id": ANY,
                    "name": "Click counter",
                    "stage": "beta",
                },
            ],
        }


class TestPreviewList(BaseTest, QueryMatchingTest):
    def setUp(self):
        cache.clear()
        super().setUp()
        # it is really important to know that /decide is CSRF exempt. Enforce checking in the client
        self.client = Client(enforce_csrf_checks=True)
        self.client.force_login(self.user)

    def _get_features(
        self,
        token=None,
        origin="http://127.0.0.1:8000",
        ip="127.0.0.1",
    ):
        return self.client.get(
            f"/api/early_access_features/",
            data={"token": token or self.team.api_token},
            HTTP_ORIGIN=origin,
            REMOTE_ADDR=ip,
        )

    @snapshot_postgres_queries
    def test_early_access_features(self):
        Person.objects.create(
            team=self.team,
            distinct_ids=["example_id"],
            properties={"email": "example@posthog.com"},
        )

        feature_flag = FeatureFlag.objects.create(
            team=self.team,
            name=f"Feature Flag for Feature Sprocket",
            key="sprocket",
            rollout_percentage=0,
            created_by=self.user,
        )
        feature_flag2 = FeatureFlag.objects.create(
            team=self.team,
            name=f"Feature Flag for Feature Sprocket",
            key="sprocket2",
            rollout_percentage=10,
            created_by=self.user,
        )
        feature = EarlyAccessFeature.objects.create(
            team=self.team,
            name="Sprocket",
            description="A fancy new sprocket.",
            stage="beta",
            feature_flag=feature_flag,
        )
        EarlyAccessFeature.objects.create(
            team=self.team,
            name="Sprocket",
            description="A fancy new sprocket.",
            stage="alpha",
            feature_flag=feature_flag2,
        )

        self.client.logout()

        with self.assertNumQueries(2):
            response = self._get_features()
            self.assertEqual(response.status_code, 200)
            self.assertEqual(response.get("access-control-allow-origin"), "http://127.0.0.1:8000")

            self.assertListEqual(
                response.json()["earlyAccessFeatures"],
                [
                    {
                        "id": str(feature.id),
                        "name": "Sprocket",
                        "description": "A fancy new sprocket.",
                        "stage": "beta",
                        "documentationUrl": "",
                        "flagKey": "sprocket",
                    }
                ],
            )

    def test_early_access_features_beta_only(self):
        Person.objects.create(
            team=self.team,
            distinct_ids=["example_id"],
            properties={"email": "example@posthog.com"},
        )

        feature_flag = FeatureFlag.objects.create(
            team=self.team,
            name=f"Feature Flag for Feature Sprocket",
            key="sprocket",
            rollout_percentage=0,
            created_by=self.user,
        )
        feature_flag2 = FeatureFlag.objects.create(
            team=self.team,
            name=f"Feature Flag for Feature Sprocket",
            key="sprocket2",
            rollout_percentage=10,
            created_by=self.user,
        )
        feature_flag3 = FeatureFlag.objects.create(
            team=self.team,
            name=f"Feature Flag for Feature Sprocket",
            key="sprocket3",
            rollout_percentage=10,
            created_by=self.user,
        )
        feature = EarlyAccessFeature.objects.create(
            team=self.team,
            name="Sprocket",
            description="A fancy new sprocket.",
            stage="beta",
            feature_flag=feature_flag,
        )
        EarlyAccessFeature.objects.create(
            team=self.team,
            name="Sprocket",
            description="A fancy new sprocket.",
            stage="alpha",
            feature_flag=feature_flag2,
        )
        EarlyAccessFeature.objects.create(
            team=self.team,
            name="Sprocket",
            description="A fancy new sprocket.",
            stage="draft",
            feature_flag=feature_flag3,
        )

        self.client.logout()

        with self.assertNumQueries(2):
            response = self._get_features()
            self.assertEqual(response.status_code, 200)
            self.assertEqual(response.get("access-control-allow-origin"), "http://127.0.0.1:8000")

            self.assertListEqual(
                response.json()["earlyAccessFeatures"],
                [
                    {
                        "id": str(feature.id),
                        "name": "Sprocket",
                        "description": "A fancy new sprocket.",
                        "stage": "beta",
                        "documentationUrl": "",
                        "flagKey": "sprocket",
                    }
                ],
            )

    def test_early_access_features_errors_out_on_random_token(self):
        self.client.logout()

        with self.assertNumQueries(1):
            response = self._get_features(token="random_token")
            self.assertEqual(response.status_code, 401)
            self.assertEqual(
                response.json()["detail"],
                "Project API key invalid. You can find your project API key in PostHog project settings.",
            )

    def test_early_access_features_errors_out_on_no_token(self):
        self.client.logout()

        with self.assertNumQueries(0):
            response = self.client.get(f"/api/early_access_features/")
            self.assertEqual(response.status_code, 401)
            self.assertEqual(
                response.json()["detail"],
                "API key not provided. You can find your project API key in PostHog project settings.",
            )
