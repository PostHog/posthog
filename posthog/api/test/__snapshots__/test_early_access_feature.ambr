# name: TestPreviewList.test_early_access_features
  '
  SELECT "posthog_team"."id",
         "posthog_team"."uuid",
         "posthog_team"."organization_id",
         "posthog_team"."api_token",
         "posthog_team"."app_urls",
         "posthog_team"."name",
         "posthog_team"."slack_incoming_webhook",
         "posthog_team"."created_at",
         "posthog_team"."updated_at",
         "posthog_team"."anonymize_ips",
         "posthog_team"."completed_snippet_onboarding",
         "posthog_team"."has_completed_onboarding_for",
         "posthog_team"."ingested_event",
         "posthog_team"."autocapture_opt_out",
         "posthog_team"."autocapture_exceptions_opt_in",
         "posthog_team"."autocapture_exceptions_errors_to_ignore",
         "posthog_team"."session_recording_opt_in",
         "posthog_team"."session_recording_sample_rate",
         "posthog_team"."session_recording_minimum_duration_milliseconds",
         "posthog_team"."session_recording_linked_flag",
         "posthog_team"."capture_console_log_opt_in",
         "posthog_team"."capture_performance_opt_in",
         "posthog_team"."surveys_opt_in",
         "posthog_team"."session_recording_version",
         "posthog_team"."signup_token",
         "posthog_team"."is_demo",
         "posthog_team"."access_control",
         "posthog_team"."week_start_day",
         "posthog_team"."inject_web_apps",
         "posthog_team"."test_account_filters",
         "posthog_team"."test_account_filters_default_checked",
         "posthog_team"."path_cleaning_filters",
         "posthog_team"."timezone",
         "posthog_team"."data_attributes",
         "posthog_team"."person_display_name_properties",
         "posthog_team"."live_events_columns",
         "posthog_team"."recording_domains",
         "posthog_team"."primary_dashboard_id",
         "posthog_team"."extra_settings",
         "posthog_team"."correlation_config",
         "posthog_team"."session_recording_retention_period_days",
         "posthog_team"."plugins_opt_in",
         "posthog_team"."opt_out_capture",
         "posthog_team"."event_names",
         "posthog_team"."event_names_with_usage",
         "posthog_team"."event_properties",
         "posthog_team"."event_properties_with_usage",
         "posthog_team"."event_properties_numerical"
  FROM "posthog_team"
  WHERE "posthog_team"."id" = 2
  LIMIT 21
  '
---
# name: TestPreviewList.test_early_access_features.1
  '
  SELECT "posthog_featureflag"."id",
         "posthog_featureflag"."key",
         "posthog_featureflag"."name",
         "posthog_featureflag"."filters",
         "posthog_featureflag"."rollout_percentage",
         "posthog_featureflag"."team_id",
         "posthog_featureflag"."created_by_id",
         "posthog_featureflag"."created_at",
         "posthog_featureflag"."deleted",
         "posthog_featureflag"."active",
         "posthog_featureflag"."rollback_conditions",
         "posthog_featureflag"."performed_rollback",
         "posthog_featureflag"."ensure_experience_continuity",
         "posthog_featureflag"."usage_dashboard_id",
         "posthog_featureflag"."has_enriched_analytics"
  FROM "posthog_featureflag"
  WHERE ("posthog_featureflag"."active"
         AND NOT "posthog_featureflag"."deleted"
         AND "posthog_featureflag"."team_id" = 2)
  '
---
# name: TestPreviewList.test_early_access_features.2
  '
  SELECT "posthog_featureflag"."id",
         "posthog_featureflag"."key",
         "posthog_featureflag"."name",
         "posthog_featureflag"."filters",
         "posthog_featureflag"."rollout_percentage",
         "posthog_featureflag"."team_id",
         "posthog_featureflag"."created_by_id",
         "posthog_featureflag"."created_at",
         "posthog_featureflag"."deleted",
         "posthog_featureflag"."active",
         "posthog_featureflag"."rollback_conditions",
         "posthog_featureflag"."performed_rollback",
         "posthog_featureflag"."ensure_experience_continuity",
         "posthog_featureflag"."usage_dashboard_id",
         "posthog_featureflag"."has_enriched_analytics"
  FROM "posthog_featureflag"
  WHERE ("posthog_featureflag"."active"
         AND NOT "posthog_featureflag"."deleted"
         AND "posthog_featureflag"."team_id" = 2)
  '
---
# name: TestPreviewList.test_early_access_features.3
  '
  SELECT "posthog_user"."id",
         "posthog_user"."password",
         "posthog_user"."last_login",
         "posthog_user"."first_name",
         "posthog_user"."last_name",
         "posthog_user"."is_staff",
         "posthog_user"."is_active",
         "posthog_user"."date_joined",
         "posthog_user"."uuid",
         "posthog_user"."current_organization_id",
         "posthog_user"."current_team_id",
         "posthog_user"."email",
         "posthog_user"."pending_email",
         "posthog_user"."temporary_token",
         "posthog_user"."distinct_id",
         "posthog_user"."is_email_verified",
         "posthog_user"."has_seen_product_intro_for",
         "posthog_user"."email_opt_in",
         "posthog_user"."partial_notification_settings",
         "posthog_user"."anonymize_data",
         "posthog_user"."toolbar_mode",
         "posthog_user"."events_column_config"
  FROM "posthog_user"
  WHERE "posthog_user"."id" = 2
  LIMIT 21
  '
---
# name: TestPreviewList.test_early_access_features.4
  '
  SELECT "posthog_team"."id",
         "posthog_team"."uuid",
         "posthog_team"."organization_id",
         "posthog_team"."api_token",
         "posthog_team"."app_urls",
         "posthog_team"."name",
         "posthog_team"."slack_incoming_webhook",
         "posthog_team"."created_at",
         "posthog_team"."updated_at",
         "posthog_team"."anonymize_ips",
         "posthog_team"."completed_snippet_onboarding",
         "posthog_team"."has_completed_onboarding_for",
         "posthog_team"."ingested_event",
         "posthog_team"."autocapture_opt_out",
         "posthog_team"."autocapture_exceptions_opt_in",
         "posthog_team"."autocapture_exceptions_errors_to_ignore",
         "posthog_team"."session_recording_opt_in",
         "posthog_team"."session_recording_sample_rate",
         "posthog_team"."session_recording_minimum_duration_milliseconds",
         "posthog_team"."session_recording_linked_flag",
         "posthog_team"."capture_console_log_opt_in",
         "posthog_team"."capture_performance_opt_in",
         "posthog_team"."surveys_opt_in",
         "posthog_team"."session_recording_version",
         "posthog_team"."signup_token",
         "posthog_team"."is_demo",
         "posthog_team"."access_control",
         "posthog_team"."week_start_day",
         "posthog_team"."inject_web_apps",
         "posthog_team"."test_account_filters",
         "posthog_team"."test_account_filters_default_checked",
         "posthog_team"."path_cleaning_filters",
         "posthog_team"."timezone",
         "posthog_team"."data_attributes",
         "posthog_team"."person_display_name_properties",
         "posthog_team"."live_events_columns",
         "posthog_team"."recording_domains",
         "posthog_team"."primary_dashboard_id",
         "posthog_team"."extra_settings",
         "posthog_team"."correlation_config",
         "posthog_team"."session_recording_retention_period_days"
  FROM "posthog_team"
  WHERE "posthog_team"."api_token" = 'token123'
  LIMIT 21 /*controller='posthog.api.early_access_feature.early_access_features',route='%5Eapi/early_access_features/%3F%28%3F%3A%5B%3F%23%5D.%2A%29%3F%24'*/
  '
---
# name: TestPreviewList.test_early_access_features.5
  '
  SELECT "posthog_earlyaccessfeature"."id",
         "posthog_earlyaccessfeature"."team_id",
         "posthog_earlyaccessfeature"."feature_flag_id",
         "posthog_earlyaccessfeature"."name",
         "posthog_earlyaccessfeature"."description",
         "posthog_earlyaccessfeature"."stage",
         "posthog_earlyaccessfeature"."documentation_url",
         "posthog_earlyaccessfeature"."created_at",
         "posthog_featureflag"."id",
         "posthog_featureflag"."key",
         "posthog_featureflag"."name",
         "posthog_featureflag"."filters",
         "posthog_featureflag"."rollout_percentage",
         "posthog_featureflag"."team_id",
         "posthog_featureflag"."created_by_id",
         "posthog_featureflag"."created_at",
         "posthog_featureflag"."deleted",
         "posthog_featureflag"."active",
         "posthog_featureflag"."rollback_conditions",
         "posthog_featureflag"."performed_rollback",
         "posthog_featureflag"."ensure_experience_continuity",
         "posthog_featureflag"."usage_dashboard_id",
         "posthog_featureflag"."has_enriched_analytics"
  FROM "posthog_earlyaccessfeature"
  LEFT OUTER JOIN "posthog_featureflag" ON ("posthog_earlyaccessfeature"."feature_flag_id" = "posthog_featureflag"."id")
  WHERE ("posthog_earlyaccessfeature"."stage" = 'beta'
         AND "posthog_earlyaccessfeature"."team_id" = 2) /*controller='posthog.api.early_access_feature.early_access_features',route='%5Eapi/early_access_features/%3F%28%3F%3A%5B%3F%23%5D.%2A%29%3F%24'*/
  '
---
